### PR TITLE
Pin tool versions with hack/go.mod

### PR DESCRIPTION
**Pull request by karlkfi**
This change centralizes the tracking of versions for tools used for development and testing. This way, the tools and all their dependencies have their checksums stored in hack/go.sum, which improves supply chain security.

Depends on: 
- https://github.com/kubernetes-sigs/kustomize/pull/5620
- https://github.com/kubernetes-sigs/kustomize/pull/5621
- https://github.com/kubernetes-sigs/kustomize/pull/5624

Go work sync & mod tidy are in a 2nd commit to make it easier to review the rest.

### CODE REVIEW

**Comment by koba1t** [comment]
@karlkfi 
I approved the below 2 PRs.
Please `rebase master` for this PR.

**Comment by karlkfi** [comment]
@koba1t Rebased and passing tests now!

File: hack/go.mod [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+go 1.21
+
+require (
+	github.com/campoy/embedmd v1.0.0
+	github.com/cli/cli v0.6.0
+	github.com/go-bindata/go-bindata/v3 v3.1.3
+	github.com/gohugoio/hugo v0.120.4
+	github.com/golangci/golangci-lint v1.51.2
+	github.com/google/addlicense v1.1.1
+	github.com/goreleaser/goreleaser v0.179.0 // https://github.com/kubernetes-sigs/kustomize/issues/4542
```
Comment (koba1t) [inline hack/go.mod:12]: @karlkfi 
We have already removed `goreleaser` from the release flow.
I think we do not necessarily have this binary.

**Comment by koba1t** [comment]
HI @karlkfi 
I think this PR almost looks good!
Thanks for your work!

So, I have one minor comment. Could you check that above?

**Comment by karlkfi** [comment]
@koba1t I removed goreleaser and rebased.

**Comment by karlkfi** [comment]
@koba1t please review

File: hack/go.mod (hunk 1 above)
Comment (karlkfi) [reply hack/go.mod]: removed

**Comment by koba1t** [comment]
Thanks!

/lgtm
/approve

**Pull request merged**